### DISCUSSION TITLE
`--zap`-like for formulae?

### DISCUSSION THREAD

**Discussion by osalbahr** [Homebrew/discussions#4476] [Everyday usage]
### Output of `brew config`

```shell
Your system is ready to brew.
develop$ brew config
HOMEBREW_VERSION: 4.0.16-6-ge62a839
ORIGIN: https://github.com/Homebrew/brew
HEAD: e62a839001faf69160965c352b66870006c2af65
Last commit: 10 hours ago
Core tap JSON: 02 May 20:59 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: quad-core 64-bit kabylake
Clang: 14.0.0 build 1400
Git: 2.37.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 12.6.5-x86_64
CLT: 14.2.0.0.1.1668646533
Xcode: N/A
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

```
~$ podman run -it alpine
/ # uname -a
Linux e5a439686b41 6.2.9-300.fc38.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Mar 30 22:32:58 UTC 2023 x86_64 Linux
/ # exit

~$ brew uninstall podman
Uninstalling /usr/local/Cellar/podman/4.5.0... (190 files, 55.9MB)

~$ brew install podman > output.txt
~$ podman run -it alpine
/ # uname -a
Linux e1cc257f3dfc 6.2.9-300.fc38.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Mar 30 22:32:58 UTC 2023 x86_64 Linux
/ # exit
```

It looks like `uninstall podman` doesn't uninstall as much as I'd expect. When I first installed `podman`, I needed to run multiple commands to install the needed VM etc. Is there a way to delete all `--formula podman` files and configurations, in a fashion similar to `--zap` for caks? I couldn't find something similar for formulae:

```
~$ brew help uninstall
Usage: brew uninstall, remove, rm [options] installed_formula|installed_cask
[...]

Uninstall a formula or cask.

  -f, --force                      Delete all installed versions of formula.
                                   Uninstall even if cask is not installed,
                                   overwrite existing files and ignore errors
                                   when removing files.
      --zap                        Remove all files associated with a cask.
                                   May remove files which are shared between
                                   applications.
      --ignore-dependencies        Don't fail uninstall, even if formula is a
                                   dependency of any installed formulae.
      --formula, --formulae        Treat all named arguments as formulae.
      --cask, --casks              Treat all named arguments as casks.
  -d, --debug                      Display any debugging information.
  -q, --quiet                      Make some output more quiet.
  -v, --verbose                    Make some output more verbose.
  -h, --help                       Show this message.
~$ 
```

<details>
<summary>I tried `brew uninstall --verbose --debug podman` if you're wondering what operations `uninstall podman`currently does:</summary>
<pre>
~$ brew uninstall --verbose --debug podman
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading podman from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading pkg-config from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading gmp from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading wget from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading libidn2 from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading gh from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading libpng from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading ncurses from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading nettle from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading pixman from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading gallery-dl from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading gdbm from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading mpdecimal from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading cmake from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading yt-dlp from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading libunistring from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading libslirp from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading neofetch from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading glib from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading rust from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading readline from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading lz4 from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading screenresolution from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading sqlite from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading xz from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading shfmt from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading libnghttp2 from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading vde from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading capstone from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading snappy from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading openssl@3 from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading ca-certificates from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading gettext from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading qemu from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading zstd from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading tcl-tk from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading shellcheck from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading libtasn1 from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading libssh from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading pcre2 from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading libevent from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading openssl@1.1 from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading pypy3 from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading unbound from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading tldr from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading libzip from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading libusb from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading python@3.11 from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading p11-kit from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::TapLoader): loading /usr/local/Homebrew/Library/Taps/null-dev/homebrew-firefox-profile-switcher/Formula/firefox-profile-switcher-connector.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading lzo from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading jpeg-turbo from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaAPILoader): loading gnutls from API
/usr/local/Homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading virtualbox
/usr/local/Homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromAPILoader): loading wine-stable
Uninstalling /usr/local/Cellar/podman/4.5.0... (190 files, 55.9MB)
rm /usr/local/bin/podman
rm /usr/local/bin/podman-mac-helper
rm /usr/local/bin/podman-remote
rm /usr/local/etc/bash_completion.d/podman
rm /usr/local/share/fish/vendor_completions.d/podman.fish
rm /usr/local/share/man/man1/podman-attach.1
rm /usr/local/share/man/man1/podman-build.1
rm /usr/local/share/man/man1/podman-commit.1
rm /usr/local/share/man/man1/podman-container-attach.1
rm /usr/local/share/man/man1/podman-container-checkpoint.1
rm /usr/local/share/man/man1/podman-container-clone.1
rm /usr/local/share/man/man1/podman-container-commit.1
rm /usr/local/share/man/man1/podman-container-cp.1
rm /usr/local/share/man/man1/podman-container-create.1
rm /usr/local/share/man/man1/podman-container-diff.1
rm /usr/local/share/man/man1/podman-container-exec.1
rm /usr/local/share/man/man1/podman-container-exists.1
rm /usr/local/share/man/man1/podman-container-export.1
rm /usr/local/share/man/man1/podman-container-init.1
rm /usr/local/share/man/man1/podman-container-inspect.1
rm /usr/local/share/man/man1/podman-container-kill.1
rm /usr/local/share/man/man1/podman-container-list.1
rm /usr/local/share/man/man1/podman-container-logs.1
rm /usr/local/share/man/man1/podman-container-pause.1
rm /usr/local/share/man/man1/podman-container-port.1
rm /usr/local/share/man/man1/podman-container-prune.1
rm /usr/local/share/man/man1/podman-container-ps.1
rm /usr/local/share/man/man1/podman-container-rename.1
rm /usr/local/share/man/man1/podman-container-restart.1
rm /usr/local/share/man/man1/podman-container-restore.1
rm /usr/local/share/man/man1/podman-container-rm.1
rm /usr/local/share/man/man1/podman-container-run.1
rm /usr/local/share/man/man1/podman-container-start.1
rm /usr/local/share/man/man1/podman-container-stats.1
rm /usr/local/share/man/man1/podman-container-stop.1
rm /usr/local/share/man/man1/podman-container-top.1
rm /usr/local/share/man/man1/podman-container-unpause.1
rm /usr/local/share/man/man1/podman-container-update.1
rm /usr/local/share/man/man1/podman-container-wait.1
rm /usr/local/share/man/man1/podman-container.1
rm /usr/local/share/man/man1/podman-cp.1
rm /usr/local/share/man/man1/podman-create.1
rm /usr/local/share/man/man1/podman-diff.1
rm /usr/local/share/man/man1/podman-events.1
rm /usr/local/share/man/man1/podman-exec.1
rm /usr/local/share/man/man1/podman-export.1
rm /usr/local/share/man/man1/podman-generate-kube.1
rm /usr/local/share/man/man1/podman-generate-spec.1
rm /usr/local/share/man/man1/podman-generate-systemd.1
rm /usr/local/share/man/man1/podman-generate.1
rm /usr/local/share/man/man1/podman-healthcheck-run.1
rm /usr/local/share/man/man1/podman-healthcheck.1
rm /usr/local/share/man/man1/podman-help.1
rm /usr/local/share/man/man1/podman-history.1
rm /usr/local/share/man/man1/podman-image-build.1
rm /usr/local/share/man/man1/podman-image-diff.1
rm /usr/local/share/man/man1/podman-image-exists.1
rm /usr/local/share/man/man1/podman-image-history.1
rm /usr/local/share/man/man1/podman-image-import.1
rm /usr/local/share/man/man1/podman-image-inspect.1
rm /usr/local/share/man/man1/podman-image-list.1
rm /usr/local/share/man/man1/podman-image-load.1
rm /usr/local/share/man/man1/podman-image-prune.1
rm /usr/local/share/man/man1/podman-image-pull.1
rm /usr/local/share/man/man1/podman-image-push.1
rm /usr/local/share/man/man1/podman-image-rm.1
rm /usr/local/share/man/man1/podman-image-save.1
rm /usr/local/share/man/man1/podman-image-scp.1
rm /usr/local/share/man/man1/podman-image-search.1
rm /usr/local/share/man/man1/podman-image-tag.1
rm /usr/local/share/man/man1/podman-image-tree.1
rm /usr/local/share/man/man1/podman-image-untag.1
rm /usr/local/share/man/man1/podman-image.1
rm /usr/local/share/man/man1/podman-images.1
rm /usr/local/share/man/man1/podman-import.1
rm /usr/local/share/man/man1/podman-info.1
rm /usr/local/share/man/man1/podman-init.1
rm /usr/local/share/man/man1/podman-inspect.1
rm /usr/local/share/man/man1/podman-kill.1
rm /usr/local/share/man/man1/podman-kube-apply.1
rm /usr/local/share/man/man1/podman-kube-down.1
rm /usr/local/share/man/man1/podman-kube-generate.1
rm /usr/local/share/man/man1/podman-kube-play.1
rm /usr/local/share/man/man1/podman-kube.1
rm /usr/local/share/man/man1/podman-load.1
rm /usr/local/share/man/man1/podman-login.1
rm /usr/local/share/man/man1/podman-logout.1
rm /usr/local/share/man/man1/podman-logs.1
rm /usr/local/share/man/man1/podman-machine-info.1
rm /usr/local/share/man/man1/podman-machine-init.1
rm /usr/local/share/man/man1/podman-machine-inspect.1
rm /usr/local/share/man/man1/podman-machine-list.1
rm /usr/local/share/man/man1/podman-machine-os-apply.1
rm /usr/local/share/man/man1/podman-machine-os.1
rm /usr/local/share/man/man1/podman-machine-rm.1
rm /usr/local/share/man/man1/podman-machine-set.1
rm /usr/local/share/man/man1/podman-machine-ssh.1
rm /usr/local/share/man/man1/podman-machine-start.1
rm /usr/local/share/man/man1/podman-machine-stop.1
rm /usr/local/share/man/man1/podman-machine.1
rm /usr/local/share/man/man1/podman-manifest-add.1
rm /usr/local/share/man/man1/podman-manifest-annotate.1
rm /usr/local/share/man/man1/podman-manifest-create.1
rm /usr/local/share/man/man1/podman-manifest-exists.1
rm /usr/local/share/man/man1/podman-manifest-inspect.1
rm /usr/local/share/man/man1/podman-manifest-push.1
rm /usr/local/share/man/man1/podman-manifest-remove.1
rm /usr/local/share/man/man1/podman-manifest-rm.1
rm /usr/local/share/man/man1/podman-manifest.1
rm /usr/local/share/man/man1/podman-network-connect.1
rm /usr/local/share/man/man1/podman-network-create.1
rm /usr/local/share/man/man1/podman-network-disconnect.1
rm /usr/local/share/man/man1/podman-network-exists.1
rm /usr/local/share/man/man1/podman-network-inspect.1
rm /usr/local/share/man/man1/podman-network-ls.1
rm /usr/local/share/man/man1/podman-network-prune.1
rm /usr/local/share/man/man1/podman-network-rm.1
rm /usr/local/share/man/man1/podman-network-update.1
rm /usr/local/share/man/man1/podman-network.1
rm /usr/local/share/man/man1/podman-pause.1
rm /usr/local/share/man/man1/podman-pod-clone.1
rm /usr/local/share/man/man1/podman-pod-create.1
rm /usr/local/share/man/man1/podman-pod-exists.1
rm /usr/local/share/man/man1/podman-pod-inspect.1
rm /usr/local/share/man/man1/podman-pod-kill.1
rm /usr/local/share/man/man1/podman-pod-logs.1
rm /usr/local/share/man/man1/podman-pod-pause.1
rm /usr/local/share/man/man1/podman-pod-prune.1
rm /usr/local/share/man/man1/podman-pod-ps.1
rm /usr/local/share/man/man1/podman-pod-restart.1
rm /usr/local/share/man/man1/podman-pod-rm.1
rm /usr/local/share/man/man1/podman-pod-start.1
rm /usr/local/share/man/man1/podman-pod-stats.1
rm /usr/local/share/man/man1/podman-pod-stop.1
rm /usr/local/share/man/man1/podman-pod-top.1
rm /usr/local/share/man/man1/podman-pod-unpause.1
rm /usr/local/share/man/man1/podman-pod.1
rm /usr/local/share/man/man1/podman-port.1
rm /usr/local/share/man/man1/podman-ps.1
rm /usr/local/share/man/man1/podman-pull.1
rm /usr/local/share/man/man1/podman-push.1
rm /usr/local/share/man/man1/podman-rename.1
rm /usr/local/share/man/man1/podman-restart.1
rm /usr/local/share/man/man1/podman-rm.1
rm /usr/local/share/man/man1/podman-rmi.1
rm /usr/local/share/man/man1/podman-run.1
rm /usr/local/share/man/man1/podman-save.1
rm /usr/local/share/man/man1/podman-search.1
rm /usr/local/share/man/man1/podman-secret-create.1
rm /usr/local/share/man/man1/podman-secret-exists.1
rm /usr/local/share/man/man1/podman-secret-inspect.1
rm /usr/local/share/man/man1/podman-secret-ls.1
rm /usr/local/share/man/man1/podman-secret-rm.1
rm /usr/local/share/man/man1/podman-secret.1
rm /usr/local/share/man/man1/podman-start.1
rm /usr/local/share/man/man1/podman-stats.1
rm /usr/local/share/man/man1/podman-stop.1
rm /usr/local/share/man/man1/podman-system-connection-add.1
rm /usr/local/share/man/man1/podman-system-connection-default.1
rm /usr/local/share/man/man1/podman-system-connection-list.1
rm /usr/local/share/man/man1/podman-system-connection-remove.1
rm /usr/local/share/man/man1/podman-system-connection-rename.1
rm /usr/local/share/man/man1/podman-system-connection.1
rm /usr/local/share/man/man1/podman-system-df.1
rm /usr/local/share/man/man1/podman-system-events.1
rm /usr/local/share/man/man1/podman-system-info.1
rm /usr/local/share/man/man1/podman-system-prune.1
rm /usr/local/share/man/man1/podman-system.1
rm /usr/local/share/man/man1/podman-tag.1
rm /usr/local/share/man/man1/podman-top.1
rm /usr/local/share/man/man1/podman-unpause.1
rm /usr/local/share/man/man1/podman-untag.1
rm /usr/local/share/man/man1/podman-update.1
rm /usr/local/share/man/man1/podman-version.1
rm /usr/local/share/man/man1/podman-volume-create.1
rm /usr/local/share/man/man1/podman-volume-exists.1
rm /usr/local/share/man/man1/podman-volume-inspect.1
rm /usr/local/share/man/man1/podman-volume-ls.1
rm /usr/local/share/man/man1/podman-volume-prune.1
rm /usr/local/share/man/man1/podman-volume-reload.1
rm /usr/local/share/man/man1/podman-volume-rm.1
rm /usr/local/share/man/man1/podman-volume.1
rm /usr/local/share/man/man1/podman-wait.1
rm /usr/local/share/man/man1/podman.1
rm /usr/local/share/zsh/site-functions/_podman
~$ 
</pre>
</details>

**Comment by osalbahr** [comment]
Surprisingly, `brew uninstall podman-desktop --zap` doesn't seem to delete all related files either. I don't know if this is something I'd need to create an issue about, since `docker`'s `--zap` does delete a lot of files including config files.

```
develop$ podman
-bash: podman: command not found

develop$ brew install podman-desktop
==> Downloading https://github.com/containers/podman-desktop/releases/download/v
Already downloaded: /Users/osos/Library/Caches/Homebrew/downloads/b7b3d798a7f159a99224522e0c79f3ef8078bb6199a7ee3c6dc7891e1feb8948--podman-desktop-0.14.1-x64.dmg
==> Installing dependencies: podman
==> Fetching podman
==> Downloading https://ghcr.io/v2/homebrew/core/podman/manifests/4.5.0
Already downloaded: /Users/osos/Library/Caches/Homebrew/downloads/498d0dc0bb747d34e099828eab216758a41ae02ddb4f657db61d84ed7fc7d753--podman-4.5.0.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/podman/blobs/sha256:fe71fa878de
Already downloaded: /Users/osos/Library/Caches/Homebrew/downloads/63963e2a1df93ee1c5b1bf60ead970926ced26fabd687ac81e51560dfbd44ad2--podman--4.5.0.monterey.bottle.tar.gz
==> Installing podman
==> Pouring podman--4.5.0.monterey.bottle.tar.gz
🍺  /usr/local/Cellar/podman/4.5.0: 190 files, 55.9MB
==> Installing Cask podman-desktop
==> Moving App 'Podman Desktop.app' to '/Applications/Podman Desktop.app'
🍺  podman-desktop was successfully installed!

develop$ podman run -it alpine
/ # uname -a
Linux 0994dd3bc8ca 6.2.9-300.fc38.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Mar 30 22:32:58 UTC 2023 x86_64 Linux
/ # exit

develop$ brew uninstall podman-desktop --zap
==> Implied `brew uninstall --cask podman-desktop`
==> Backing App 'Podman Desktop.app' up to '/usr/local/Caskroom/podman-desktop/0
==> Removing App '/Applications/Podman Desktop.app'
==> Dispatching zap stanza
==> Trashing files:
~/.local/share/containers/podman-desktop
==> Removing all staged versions of Cask 'podman-desktop'

develop$ podman run -it alpine
/ # uname -a
Linux 00d5e4ddd38c 6.2.9-300.fc38.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Mar 30 22:32:58 UTC 2023 x86_64 Linux
/ # exit

develop$ brew doctor
Your system is ready to brew.
```

Note: the space between commands is only for display purposes. I ran them all sequentially, after uninstalling `podman`.

**Comment by SMillerDev** [comment]
No, brew only uninstalls the files it installed. Not the files the program generated.

**Reply by osalbahr** [reply]
I know that this feature doesn’t exist for formulae. I’m wondering if it would be a good idea to add it as an option, similarly to `--zap`.
